File: SortVision/public/code/heap/java/heapSort.java
public class HeapSort {
    // Time: Best/Avg/Worst O(n log n), Space: O(1)
    public static void heapSort(int[] arr) {
        if (arr == null || arr.length <= 1) return;

        int n = arr.length;
        for (int i = n / 2 - 1; i >= 0; i--) heapify(arr, n, i);
        for (int i = n - 1; i > 0; i--) {
            int t = arr[0];
            arr[0] = arr[i];
            arr[i] = t;
            heapify(arr, i, 0);
        }
    }

    private static void heapify(int[] arr, int n, int i) {
        int largest = i;
        int left = 2 * i + 1;
        int right = 2 * i + 2;

        if (left < n && arr[left] > arr[largest]) largest = left;
        if (right < n && arr[right] > arr[largest]) largest = right;

        if (largest != i) {
            int t = arr[i];
            arr[i] = arr[largest];
            arr[largest] = t;
            heapify(arr, n, largest);
        }
    }

    public static void main(String[] args) {
        int[] arr = {64, 25, 12, 22, 11};
        heapSort(arr);
        for (int x : arr) System.out.print(x + " ");
        System.out.println();
    }
}

File: SortVision/public/code/heap/kotlin/heapSort.kt
fun heapSort(arr: IntArray): IntArray {
    // Time: Best/Avg/Worst O(n log n), Space: O(1)
    val n = arr.size

    for (i in n / 2 - 1 downTo 0) heapify(arr, n, i)
    for (i in n - 1 downTo 1) {
        val t = arr[0]
        arr[0] = arr[i]
        arr[i] = t
        heapify(arr, i, 0)
    }
    return arr
}

private fun heapify(arr: IntArray, n: Int, i: Int) {
    var largest = i
    val left = 2 * i + 1
    val right = 2 * i + 2

    if (left < n && arr[left] > arr[largest]) largest = left
    if (right < n && arr[right] > arr[largest]) largest = right

    if (largest != i) {
        val t = arr[i]
        arr[i] = arr[largest]
        arr[largest] = t
        heapify(arr, n, largest)
    }
}

fun main() {
    val arr = intArrayOf(64, 25, 12, 22, 11)
    println(heapSort(arr).joinToString())
}
